Signal/src/util/OWSBackupExport.m

```diff
@@ -0,0 +1,601 @@
+@property (nonatomic, nullable) NSString *manifestRecordName;
+
+@property (nonatomic) NSString *exportDirPath;
+
+@end
+
+#pragma mark -
+
+@implementation OWSBackupExport
+
+- (instancetype)initWithDelegate:(id<OWSBackupExportDelegate>)delegate
+                  primaryStorage:(OWSPrimaryStorage *)primaryStorage
+{
+    self = [super init];
+
+    if (!self) {
+        return self;
+    }
+
+    OWSAssert(primaryStorage);
+    OWSAssert([OWSStorage isStorageReady]);
+
+    self.delegate = delegate;
+    _srcDBConnection = primaryStorage.newDatabaseConnection;
+
+    return self;
+}
+
+- (void)dealloc
+{
+    // Surface memory leaks by logging the deallocation.
+    DDLogVerbose(@"Dealloc: %@", self.class);
+
+    [[NSNotificationCenter defaultCenter] removeObserver:self];
+
+    if (self.exportDirPath) {
+        [OWSFileSystem deleteFileIfExists:self.exportDirPath];
+    }
+}
+
+- (void)startAsync
+{
+    OWSAssertIsOnMainThread();
+
+    DDLogInfo(@"%@ %s", self.logTag, __PRETTY_FUNCTION__);
+
+    self.backgroundTask = [OWSBackgroundTask backgroundTaskWithLabelStr:__PRETTY_FUNCTION__];
+
+    __weak OWSBackupExport *weakSelf = self;
+    [OWSBackupAPI checkCloudKitAccessWithCompletion:^(BOOL hasAccess) {
+        if (hasAccess) {
+            dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+                [weakSelf start];
+            });
+        }
+    }];
+}
+
+- (void)start
+{
+    __weak OWSBackupExport *weakSelf = self;
+    [self configureExport:^(BOOL success) {
+        if (!success) {
+            [self failWithErrorDescription:
+                      NSLocalizedString(@"BACKUP_EXPORT_ERROR_COULD_NOT_EXPORT",
+                          @"Error indicating the a backup export could not export the user's data.")];
+            return;
+        }
+
+        if (self.isComplete) {
+            return;
+        }
+        if (![self exportDatabase]) {
+            [self failWithErrorDescription:
+                      NSLocalizedString(@"BACKUP_EXPORT_ERROR_COULD_NOT_EXPORT",
+                          @"Error indicating the a backup export could not export the user's data.")];
+            return;
+        }
+        if (self.isComplete) {
+            return;
+        }
+        [self saveToCloud:^(NSError *_Nullable error) {
+            if (error) {
+                [weakSelf failWithError:error];
+            } else {
+                [weakSelf succeed];
+            }
+        }];
+    }];
+}
+
+- (void)configureExport:(OWSBackupExportBoolCompletion)completion
+{
+    DDLogVerbose(@"%@ %s", self.logTag, __PRETTY_FUNCTION__);
+
+    NSString *temporaryDirectory = NSTemporaryDirectory();
+    self.exportDirPath = [temporaryDirectory stringByAppendingString:[NSUUID UUID].UUIDString];
+    NSString *exportDatabaseDirPath = [self.exportDirPath stringByAppendingPathComponent:@"Database"];
+    self.databaseSalt = [Randomness generateRandomBytes:(int)kSQLCipherSaltLength];
```

Comment:
I think it'll be nice to use the "backup key" retrieved from the service as the "SQL Cipher password" for the exported database image.  We use a random salt that I'll eventually include in the backup manifest.

Reply:
Alternately we could just use a random keyspec and store that in the manifest.  Maybe that's simpler.  Yeah.